### ISSUE TITLE
Status bar colors not really an issue... more of enhancement/discussion

### ISSUE THREAD

**Issue by cyansprite** [onivim/oni#604]
I was playing around tonight... I'm pretty picky when it comes to colors.

<a href="http://imgur.com/MgS5vjb"><img src="http://i.imgur.com/MgS5vjb.gif" title="source: imgur.com" /></a>

achieved by (oni statusbar core), but my concerned is... should we have a color file, or should it just be part of the config file? It's what I ran into while playing around with the keybindings...

```js
        const getBackgroundColorForMode = (m) => {
            switch (m) {
                case "insert":
                    return Oni.configuration.getValue("statusbar.insertBackgroundColor", "rgb(0,200,100)")
                case "visual":
                    return Oni.configuration.getValue("statusbar.visualBackgroundColor", "rgb(0,100,255)")
                case "replace":
                    return Oni.configuration.getValue("statusbar.replaceBackgroundColor", "rgb(0,200,100)")
                case "operator":
                    return Oni.configuration.getValue("statusbar.operatorBackgroundColor", "rgb(255,100,0)")
                default:
                    return Oni.configuration.getValue("statusbar.normalBackgroundColor", "rgb(0,100,255)")
            }
        }
        const getForegroundColorForMode = (m) => {
            switch (m) {
                case "insert":
                    return Oni.configuration.getValue("statusbar.insertForegroundColor", "rgb(220, 220, 220)")
                case "visual":
                    return Oni.configuration.getValue("statusbar.visualForegroundColor", "rgb(220, 220, 220)")
                case "replace":
                    return Oni.configuration.getValue("statusbar.replaceForegroundColor", "rgb(220, 220, 220)")
                case "operator":
                    return Oni.configuration.getValue("statusbar.operatorForegroundColor", "rgb(220, 220, 220)")
                default:
                    return Oni.configuration.getValue("statusbar.normalForegroundColor", "rgb(220, 220, 220)")
            }
        }

```

**Comment by CrossR** [comment]
I remember poking around the tickets and seeing this: https://github.com/extr0py/oni/issues/33

Obviously no idea if it changed since then, but it looks like goal at one point was to have a separate file for it.

I think that makes more sense, as potentially you'd want to share colourschemes etc, without having to remove your config from it too.

**Comment by cyansprite** [comment]
Yeah, I agree it should be a separate file...
I'm not sure if it should be a .js or .tmTheme, or .json, or .less...
However, abstracting Config.ts would be the best option for .js file, it would match the current format of the Config.

Of course we could change this altogether, json is super light weight to parse.
.less could be implemented natively and so we don't have to parse it really.
.tmTheme I've never really looked at.

**Comment by extr0py** [comment]
> I remember poking around the tickets and seeing this: #33

Definitely, this has been an issue for a while! And even moreso now that there are more extensibility points in the UI, that aren't covered by vim colorschemes.

For now, exposing customizations in the configuration seems reasonable. 

I'm thinking something like:
```
{
   "oni.colors": {
           "statusbar.background": "black",
           "statusbar.mode.normal.background": "red",
           "quickInfo.background": "#000000",
          // ...etc
     } 
}
```
(Which is pretty similiar to what you suggested above - thanks @cyansprite !)

That seems like a reasonable first step - if someone _really_ doesn't like the default colors, they can at least override them in their configuration!

In terms of a next step, I'd like to be able to bundle the following:
- vim themes
- oni-specific styling
- CSS (for animations)

There are a few things we'd need to add to make this happen:
- __Augment the package.json format__ with the name of the colorscheme, included, plus dependent vim colorschemes.... something like this:
```
        "colorSchemes": [{
            "type": "dark",
            "colors": {
                "statusbar.background": "#FF0000",
                "quickinfo.background": "#00FF00"
            },
            "tokenColors": "./colors.vim",
            "vimSchemeName": "darkblue",
            "css": "./animations.css"
        }]
```

Once we have that set, we'd want to hook the `ColorScheme` autocmd from neovim. When we see the scheme defined by `vimSchemeName`, we'd apply the colors / css / etc on the Oni side. Any overrides in `config.js` would take precedence here, though.

Further on from there, it'd be nice to be able to _fully define_ a color scheme in JS/JSON, so that the `tokenColors` (vim colors file) isn't needed. But that isn't necessary at the start.

Another cool thing would be to list all the available colorschemes in the app menu.

Anyway that's just a brain-dump of what I'm thinking in terms of having custom color schemes... hopefully that was coherent 😄 Sounds pretty in-line with what was discussed above. I like being able to quickly tweak colors in the config.js, and then extract them out into a separate color-theme plugin - so I think we'd be able to accomplish that here.

**Comment by cyansprite** [comment]
I will implement something like this and let you know how it flows with a branch and PR of course :P.
```{
   "oni.colors": {
           "statusbar.background": "black",
           "statusbar.mode.normal.background": "red",
           "quickInfo.background": "#000000",
          // ...etc
     } 
}
```
I will also work on a menu item and picking a color scheme... maybe we can make the vim-colorscheme plugin something we can use ( Do we support term colors? )
If not the actual plugin we could wrap it into a js plugin :P.

**Comment by cyansprite** [comment]
Also a feauture I would want to add is cursorline changing color in insertmode, its a feature I like.

**Comment by extr0py** [comment]
> Also a feauture I would want to add is cursorline changing color in insertmode, its a feature I like.

Excellent, that definitely seems doable, if we have colors defined for the cursorline / cursorcolumn.

This kind of goes with #610 - if we expose an `activate` method in `config.js`, we could set up stuff like that as a personalized configuration.

**Comment by bryphe** [comment]
This is actually implemented now with #975 - need to add some documentation.

An example file using the theme is here:
https://github.com/onivim/oni/blob/master/extensions/theme-onedark/colors/onedark.json

And all the colors can now be overridden in the `config.js`, like:

```
"colors.highlight.mode.normal.background": "yellow",
```

**Comment by bryphe** [comment]
Closing, some further work tracked in #412 (the `css` file portion), and it'd be nice to also have `tokenColors` specified in these themes.
